src/pip/_internal/models/link.py

```diff
@@ -33,13 +33,12 @@ class Link(KeyBasedCompareMixin):
 
     def __init__(
         self,
-        url,                   # type: str
-        comes_from=None,       # type: Optional[Union[str, HTMLPage]]
-        requires_python=None,  # type: Optional[str]
-        yanked_reason=None,    # type: Optional[str]
-        cache_link_parsing=True,  # type: bool
-    ):
-        # type: (...) -> None
+        url: str,
+        comes_from: Optional[Union[HTMLPage]] = None,
```

Comment:
I think the issue is here. Changing the `comes_from` annotation from:

```python
comes_from: Optional[Union[HTMLPage]] = None,
```

to this one:

```python
comes_from: Optional[HTMLPage] = None,
```

must solve the problem (like Mypy suggested).

Reply:
I'm still not sure of this, but I don't see any other option. I will make it.

Reply:
Hmmm... It failed again. Maybe something bigger is wrong at there.

Reply:
It still failing. Or maybe I can use:

```python
Optional[Any] = None,
```

Reply:
`comes_from` is currently `Optional[Union[str, HTMLPage]]`, which looks correct to me?

Reply:
Oh, seems like I made a (really simple) mistake which is generating the mypy problem :man_facepalming:\: I will fix it on a commit. Thanks @uranusjr.

Reply:
It fails again. I think it is related to these lines of code:

```python
if TYPE_CHECKING:
    from pip._internal.index.collector import HTMLPage
```

If we remove the `if` statement, I think it may help. What do you think?

Reply:
No I think that's to prevent a circular import issue. Just turn `HTMLPage` into a str in the annotation.

Reply:
Ok.